File: packages/runtime-protocol/src/terminal-multiplex/connection-records.ts
export const TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE = {
  phase: { offer: 0, accept: 1 },
  appState: { foreground: 0, background: 1, unknown: 2 },
  epoch: {
    bytes: 24,
    phaseOffset: 0,
    protocolMinorOffset: 1,
    reserved16Offset: 2,
    maxFrameBytesOffset: 4,
    maxStreamsOffset: 8,
    heartbeatMsOffset: 12,
    connectionGenerationOffset: 16,
    reserved32Offset: 20
  },
  heartbeat: {
    bytes: 16,
    phaseOffset: 0,
    appStateOffset: 1,
    reserved16Offset: 2,
    senderQueueBytesOffset: 4,
    monotonicMicrosOffset: 8
  }
} as const

type TerminalMultiplexHandshakePhase =
  (typeof TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase)[keyof typeof TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase]

type TerminalMultiplexHeartbeatAppState =
  (typeof TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState)[keyof typeof TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState]

export type TerminalMultiplexEpochRecord = {
  phase: TerminalMultiplexHandshakePhase
  protocolMinor: number
  maxFrameBytes: number
  maxStreams: number
  heartbeatMs: number
  connectionGeneration: number
}

export type TerminalMultiplexHeartbeatRecord = {
  phase: TerminalMultiplexHandshakePhase
  appState: TerminalMultiplexHeartbeatAppState
  senderQueueBytes: number
  monotonicMicros: bigint
}

export function encodeTerminalMultiplexEpochRecord(
  record: TerminalMultiplexEpochRecord
): Uint8Array<ArrayBuffer> {
  const wire = TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.epoch
  const out = new Uint8Array(wire.bytes)
  const view = new DataView(out.buffer)
  view.setUint8(wire.phaseOffset, record.phase)
  view.setUint8(wire.protocolMinorOffset, record.protocolMinor)
  view.setUint16(wire.reserved16Offset, 0, true)
  view.setUint32(wire.maxFrameBytesOffset, record.maxFrameBytes, true)
  view.setUint32(wire.maxStreamsOffset, record.maxStreams, true)
  view.setUint32(wire.heartbeatMsOffset, record.heartbeatMs, true)
  view.setUint32(wire.connectionGenerationOffset, record.connectionGeneration, true)
  view.setUint32(wire.reserved32Offset, 0, true)
  return out
}

export function decodeTerminalMultiplexEpochRecord(
  payload: Uint8Array<ArrayBufferLike>
): TerminalMultiplexEpochRecord | null {
  const wire = TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.epoch
  if (payload.byteLength !== wire.bytes) {
    return null
  }
  const view = payloadView(payload)
  const phase = view.getUint8(wire.phaseOffset)
  if (
    (phase !== TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase.offer &&
      phase !== TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase.accept) ||
    view.getUint16(wire.reserved16Offset, true) !== 0 ||
    view.getUint32(wire.reserved32Offset, true) !== 0
  ) {
    return null
  }
  return {
    phase,
    protocolMinor: view.getUint8(wire.protocolMinorOffset),
    maxFrameBytes: view.getUint32(wire.maxFrameBytesOffset, true),
    maxStreams: view.getUint32(wire.maxStreamsOffset, true),
    heartbeatMs: view.getUint32(wire.heartbeatMsOffset, true),
    connectionGeneration: view.getUint32(wire.connectionGenerationOffset, true)
  }
}

export function encodeTerminalMultiplexHeartbeatRecord(
  record: TerminalMultiplexHeartbeatRecord
): Uint8Array<ArrayBuffer> {
  const wire = TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.heartbeat
  const out = new Uint8Array(wire.bytes)
  const view = new DataView(out.buffer)
  view.setUint8(wire.phaseOffset, record.phase)
  view.setUint8(wire.appStateOffset, record.appState)
  view.setUint16(wire.reserved16Offset, 0, true)
  view.setUint32(wire.senderQueueBytesOffset, record.senderQueueBytes, true)
  view.setBigUint64(wire.monotonicMicrosOffset, record.monotonicMicros, true)
  return out
}

export function decodeTerminalMultiplexHeartbeatRecord(
  payload: Uint8Array<ArrayBufferLike>
): TerminalMultiplexHeartbeatRecord | null {
  const wire = TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.heartbeat
  if (payload.byteLength !== wire.bytes) {
    return null
  }
  const view = payloadView(payload)
  const phase = view.getUint8(wire.phaseOffset)
  const appState = view.getUint8(wire.appStateOffset)
  if (
    (phase !== TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase.offer &&
      phase !== TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.phase.accept) ||
    appState > TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.unknown ||
    view.getUint16(wire.reserved16Offset, true) !== 0
  ) {
    return null
  }
  return {
    phase,
    appState: decodeHeartbeatAppState(appState),
    senderQueueBytes: view.getUint32(wire.senderQueueBytesOffset, true),
    monotonicMicros: view.getBigUint64(wire.monotonicMicrosOffset, true)
  }
}

function decodeHeartbeatAppState(appState: number): TerminalMultiplexHeartbeatAppState {
  if (appState === TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.foreground) {
    return TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.foreground
  }
  if (appState === TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.background) {
    return TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.background
  }
  return TERMINAL_MULTIPLEX_CONNECTION_RECORD_WIRE.appState.unknown
}

function payloadView(payload: Uint8Array<ArrayBufferLike>): DataView {
  return new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
}

File: apps/mobile-ios/YiruMobile/Platform/Terminal/Multiplex/ConnectionRecords.swift
import Foundation

nonisolated struct TerminalMultiplexEpochRecord: Equatable, Sendable {
    let phase: TerminalMultiplexHandshakePhase
    let protocolMinor: UInt8
    let maxFrameBytes: UInt32
    let maxStreams: UInt32
    let heartbeatMilliseconds: UInt32
    let connectionGeneration: UInt32
}

nonisolated struct TerminalMultiplexHeartbeatRecord: Equatable, Sendable {
    let phase: TerminalMultiplexHandshakePhase
    let appState: TerminalMultiplexAppState
    let senderQueueBytes: UInt32
    let monotonicMicroseconds: UInt64
}

nonisolated enum TerminalMultiplexConnectionRecordCodec {
    static func encode(_ record: TerminalMultiplexEpochRecord) -> Data {
        let wire = TerminalMultiplexEpochRecordWire.self
        var data = Data(repeating: 0, count: wire.bytes)
        data[wire.phaseOffset] = record.phase.rawValue
        data[wire.protocolMinorOffset] = record.protocolMinor
        TerminalWireBytes.write(record.maxFrameBytes, to: &data, at: wire.maxFrameBytesOffset)
        TerminalWireBytes.write(record.maxStreams, to: &data, at: wire.maxStreamsOffset)
        TerminalWireBytes.write(
            record.heartbeatMilliseconds,
            to: &data,
            at: wire.heartbeatMsOffset
        )
        TerminalWireBytes.write(
            record.connectionGeneration,
            to: &data,
            at: wire.connectionGenerationOffset
        )
        return data
    }

    static func decodeEpoch(_ data: Data) -> TerminalMultiplexEpochRecord? {
        let wire = TerminalMultiplexEpochRecordWire.self
        guard data.count == wire.bytes,
            let phase = TerminalMultiplexHandshakePhase(
                rawValue: TerminalWireBytes.byte(in: data, at: wire.phaseOffset)
            ),
            TerminalWireBytes.uint16(in: data, at: wire.reserved16Offset) == 0,
            TerminalWireBytes.uint32(in: data, at: wire.reserved32Offset) == 0
        else {
            return nil
        }
        return TerminalMultiplexEpochRecord(
            phase: phase,
            protocolMinor: TerminalWireBytes.byte(in: data, at: wire.protocolMinorOffset),
            maxFrameBytes: TerminalWireBytes.uint32(in: data, at: wire.maxFrameBytesOffset),
            maxStreams: TerminalWireBytes.uint32(in: data, at: wire.maxStreamsOffset),
            heartbeatMilliseconds: TerminalWireBytes.uint32(
                in: data,
                at: wire.heartbeatMsOffset
            ),
            connectionGeneration: TerminalWireBytes.uint32(
                in: data,
                at: wire.connectionGenerationOffset
            )
        )
    }

    static func encode(_ record: TerminalMultiplexHeartbeatRecord) -> Data {
        let wire = TerminalMultiplexHeartbeatRecordWire.self
        var data = Data(repeating: 0, count: wire.bytes)
        data[wire.phaseOffset] = record.phase.rawValue
        data[wire.appStateOffset] = record.appState.rawValue
        TerminalWireBytes.write(
            record.senderQueueBytes,
            to: &data,
            at: wire.senderQueueBytesOffset
        )
        TerminalWireBytes.write(
            record.monotonicMicroseconds,
            to: &data,
            at: wire.monotonicMicrosOffset
        )
        return data
    }

    static func decodeHeartbeat(_ data: Data) -> TerminalMultiplexHeartbeatRecord? {
        let wire = TerminalMultiplexHeartbeatRecordWire.self
        guard data.count == wire.bytes,
            let phase = TerminalMultiplexHandshakePhase(
                rawValue: TerminalWireBytes.byte(in: data, at: wire.phaseOffset)
            ),
            let appState = TerminalMultiplexAppState(
                rawValue: TerminalWireBytes.byte(in: data, at: wire.appStateOffset)
            ),
            TerminalWireBytes.uint16(in: data, at: wire.reserved16Offset) == 0
        else {
            return nil
        }
        return TerminalMultiplexHeartbeatRecord(
            phase: phase,
            appState: appState,
            senderQueueBytes: TerminalWireBytes.uint32(
                in: data,
                at: wire.senderQueueBytesOffset
            ),
            monotonicMicroseconds: TerminalWireBytes.uint64(
                in: data,
                at: wire.monotonicMicrosOffset
            )
        )
    }
}
